### DISCUSSION TITLE
Lua Helper to configure vapoursynth plugin path on python side

### DISCUSSION THREAD

**Discussion by sosie-js** [arch1t3cht/Aegisub#147] [Show and tell]
Hello arch1t3cht,

After patching aegisub-vs.py by commenting it like this

```
    '''
    if b"-Dcachedir" not in core.lsmas.Version()["config"]: # type: ignore
        raise vs.Error("To use Aegisub's LWLibavSource wrapper, the `lsmas` plugin must support the `cachedir` option for LWLibavSource.")
    '''
```

due to error
```
     if b"-Dcachedir" not in core.lsmas.Version()["config"]: # type: ignore
    File "src/cython/vapoursynth.pyx", line 2792, in vapoursynth.Plugin.__getattr__
    AttributeError: There is no function named Version
```

I succeed to make vapoursynth work in linux.  As this file is embedded into Appimage  it cannot be changed unless you unpack 
and repack appimage. 

 I am wondering if a python module can be used instead so everyone can improve it and contribute using
a fork of [ your later release](https://github.com/arch1t3cht/Aegisub/blob/vapoursynth/automation/vapoursynth/aegisub_vs.py)

I started [peagisub](https://github.com/sosie-js/peagisub-vs/) to collect vapoursynth and aegisub paths in lua and expose those as config file so python can read back to set plugin path , what do you think of [starting it on python side](https://github.com/luarocks/luarocks/issues/1694) and adding it to aegisub_vs.py script?

**Comment by arch1t3cht** [comment]
The `Version` `AttributeError` was due to the version detection method not working for certain lsmas forks. I had a fix for that already which I just forgot to push, so I pushed it now.
> I succeed to make vapoursynth work in linux. As this file is embedded into Appimage it cannot be changed unless you unpack
and repack appimage.
> 
> I am wondering if a python module can be used instead so everyone can improve it and contribute using
a fork of[ your later release](https://github.com/arch1t3cht/Aegisub/blob/vapoursynth/automation/vapoursynth/aegisub_vs.py)

The envisioned way to make custom changes to this python module is to copy it into your user-specific Aegisub directory. On Linux you'd copy it to, say, `~/.aegisub/automation/vapoursynth/aegisub_vs_custom.py`, make your desired changes, and update your default script in the settings to import from this module instead. The directory is added to the python import path. Of course you can also use a mixed system where you import some functions from the official module and others from your custom one.

**Reply by sosie-js** [reply]
Thanks a lot for this very clear explanation Arch1t3cht.
 I have quite the same idea but I did not realize the default script was the entry point to import aegisub_vs.py,
However the default script is not triggered when vps scripts are as input videos which is my case so i have to hook your aegisub_vs.py requiring unpack/repack no?

The thing I want in fact discuss was the line "a.set_paths(locals())" in the default script.
It will not do what needed unless you have set an env I don't see for now. This is the reason why I built the lua 
module peagisub. Since I wondered if we can not do better. 

Indeed,  I saw since you installed on windows the vapoursynth plugins into ?data/vapoursynth at least 5 in the
non portable version
```
BestAudioSource.dll  libscxvid.dll          libwwxd64.dll
BestSource.dll       libvslsmashsource.dll
```
less in the portable version
```
BestAudioSource.dll
libscxvid.dll
libvslsmashsource.dll
libwwxd64.dll
```
these are not in the Ubuntu image but i succeed to produce the plugins and autoload them once the path are set

The main problem is that for now configuration of vapoursynth is not very normalized, I hope in the future that we can find agreement in the vapoursynth.conf to make 
the UserpluginDIr path standard be ?user/automation 
the SystemPluginDir $prefix/vapoursynth with $prefix=/usr/local on linux avoiding env gaz factories

It worth noticing then than just setting a magick link from python's lib into python's lib-dynload like svp-team pointed out in [Vapoursynth Building](https://www.svp-team.com/wiki/SVP:Linux#Common_packages) makes vspipe working
```
# make Python find the Vapoursynth module
# replace "python3.10" with the actual version you have - for example, "python3.12"
sudo ln -s /usr/local/lib/python3.10/site-packages/vapoursynth.so /usr/lib/python3.10/lib-dynload/vapoursynth.so
```
I have build on my side a python package I intend to submit to the vapoursynth / bestsource author . I have sound in vspreview now but not yet in aegisub when using vapoursynth.

Cheers,
                           SoSie

**Reply by sosie-js** [reply]
I get the audio working with the latest release
I never succeed to compile bas plugin in linux, better switch imho  to [bestsource as suggested by its author](http://www.vapoursynth.com/2024/05/bestsource-2-frame-accurate-seeking-in-everything/)
```
# This default script will load an audio file using BestSource.
# It requires the `bs` plugin.

import vapoursynth as vs
import aegisub_vs as a
a.set_paths(locals())

a.ensure_plugin("bs", "BestSource", "To use Aegisub's default audio loader, the `bs` plugin for VapourSynth must be installed")
vs.core.bs.AudioSource(source=filename).set_output()
``` 

Normally if we follow the logic of vapoursynth's author, video is set_output as channel 0 and audio channel 1 .
Here what would be good is to have a var indicating which channel we need as this  dump using pprint give no indication - I can confim now the default user script  is not used -
``` 
from pprint import pprint
with open( "dump_locals.txt" , "w" ) as write:
    pprint(locals(), indent=2,stream=write)
``` 
or add a compatibility to a sort of Audiodub() that mux the two channels like [suggested by the combination example](http://www.vapoursynth.com/doc/gettingstarted.html#example-script). In other words video is the first channel and audio the last.

**Reply by sosie-js** [reply]
> The `Version` `AttributeError` was due to the version detection method not working for certain lsmas forks. I had a fix for that already which I just forgot to push, so I pushed it now.

I got it thanks
https://github.com/TypesettingTools/Aegisub/commit/c49b604172f97f5dd098cecc3d08be2515d4f1ff

**Reply by sosie-js** [reply]
Hi Arch1t3cht,

I am pleased to inform you that now what I said [is working](https://github.com/sosie-js/peagisub-vs/blob/1.5.0/test/aegisub-vs.py) it uses an advanced lua bridge [https://github.com/sosie-js/peagisub-vs](peagisub) based on luarocks, the lua package manager called ![peagisub](https://github.com/sosie-js/peagisub-vs/blob/1.5.0/peagisub.png#)
I tested it on ubuntu 22.04.

**Reply by sosie-js** [reply]
@arch1t3cht 

I released also python-bestsource that is an attempt [to make trim work with audio in vapoursynth](https://github.com/sosie-js/python-bestsource?tab=readme-ov-file#aegisub-audio-tests-support-in-vapoursynth)
